keanu-project/src/main/java/io/improbable/keanu/algorithms/mcmc/tensor/TensorNUTS.java

```diff
@@ -0,0 +1,531 @@
+        );
+
+        final double logOfMasterPAfterLeapfrog = getLogProb(probabilisticVertices);
+
+        final double logOfMasterPMinusMomentum = logOfMasterPAfterLeapfrog - 0.5 * dotProduct(leapfrog.momentum);
+        final int acceptedLeapfrogCount = u <= Math.exp(logOfMasterPMinusMomentum) ? 1 : 0;
+        final boolean shouldContinueFlag = u < Math.exp(DELTA_MAX + logOfMasterPMinusMomentum);
+
+        final Map<Long, ?> sampleAtAcceptedPosition = takeSample(sampleFromVertices);
+
+        return new BuiltTree(
+            leapfrog.position,
+            leapfrog.gradient,
+            leapfrog.momentum,
+            leapfrog.position,
+            leapfrog.gradient,
+            leapfrog.momentum,
+            leapfrog.position,
+            leapfrog.gradient,
+            logOfMasterPAfterLeapfrog,
+            sampleAtAcceptedPosition,
+            acceptedLeapfrogCount,
+            shouldContinueFlag
+        );
+    }
+
+    private static double getLogProb(List<Vertex> probabilisticVertices) {
+        double sum = 0.0;
+        for (Vertex<?> vertex : probabilisticVertices) {
+            sum += vertex.logProbAtValue();
+        }
+        return sum;
+    }
+
+    private static void acceptOtherPositionWithProbability(double probability,
+                                                           BuiltTree tree,
+                                                           BuiltTree otherTree,
+                                                           KeanuRandom random) {
+        if (withProbability(probability, random)) {
+            tree.acceptedPosition = otherTree.acceptedPosition;
+            tree.gradientAtAcceptedPosition = otherTree.gradientAtAcceptedPosition;
+            tree.logOfMasterPAtAcceptedPosition = otherTree.logOfMasterPAtAcceptedPosition;
+            tree.sampleAtAcceptedPosition = otherTree.sampleAtAcceptedPosition;
+        }
+    }
+
+    private static boolean withProbability(double probability, KeanuRandom random) {
+        return random.nextDouble() < probability;
+    }
+
+    private static boolean isNotUTurning(Map<Long, DoubleTensor> positionForward,
+                                         Map<Long, DoubleTensor> positionBackward,
+                                         Map<Long, DoubleTensor> momentumForward,
+                                         Map<Long, DoubleTensor> momentumBackward) {
+        double forward = 0.0;
+        double backward = 0.0;
+
+        for (Map.Entry<Long, DoubleTensor> forwardPositionForLatent : positionForward.entrySet()) {
+
+            final long latentId = forwardPositionForLatent.getKey();
+            final DoubleTensor forwardMinusBackward = forwardPositionForLatent.getValue().minus(
+                positionBackward.get(latentId)
+            );
+
+            forward += forwardMinusBackward.times(momentumForward.get(latentId)).sum();
+            backward += forwardMinusBackward.times(momentumBackward.get(latentId)).sum();
```

Comment:
We can to timesInPlace for the 2nd times right? Not using the forwardMinusBackward tensor anymore looks like.

Reply:
Aye - great spot. Have changed to `backward += forwardMinusBackward.timesInPlace(momentumBackward.get(latentId)).sum();`